apis/v1alpha1/shared_types.go

```diff
@@ -0,0 +1,184 @@
+	// NamespacedPods defines a way to select a set of pods in
+	// in a set of namespaces.
+	// +optional
+	Pods *NamespacedPodPeer `json:"pods,omitempty"`
+}
+
+type NamespaceRelation string
+
+const (
+	NamespaceSame    NamespaceRelation = "Self"
+	NamespaceNotSame NamespaceRelation = "NotSelf"
+)
+
+// NamespacedPeer defines a flexible way to select Namespaces in a cluster.
+// Exactly one of the selectors must be set.  If a consumer observes none of
+// its fields are set, they must assume an unknown option has been specified
+// and fail closed.
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type NamespacedPeer struct {
+	// Related provides a mechanism for selecting namespaces. It can be set to
+	// "Self" to select all pods in the subject's namespace, and set to "NotSelf"
+	// to select all pods not in the subject's namespace.
+	// +optional
+	Related *NamespaceRelation `json:"related,omitempty"`
+
+	// Selector is a labelSelector used to select Namespaces, This field
+	// follows standard label selector semantics; if present but empty, it selects
+	// all Namespaces.
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// SameLabels is used to select a set of Namespaces that share the same values
+	// for a set of labels.
+	// To be selected a Namespace must have all of the labels defined in SameLabels,
+	// and they must all have the same value as the subject of this policy.
+	// If Samelabels is Empty then nothing is selected.
+	// +optional
+	// +kubebuilder:validation:MaxItems=100
+	SameLabels []string `json:"sameLabels,omitempty"`
+
+	// NotSameLabels is used to select a set of Namespaces that do not have a set
+	// of label(s). To be selected a Namespace must have none of the labels defined
+	// in NotSameLabels. If NotSameLabels is empty then nothing is selected.
+	// +optional
+	// +kubebuilder:validation:MaxItems=100
+	NotSameLabels []string `json:"notSameLabels,omitempty"`
+}
+
+// PodPeer defines a flexible way to select pods in a cluster. Exactly one of the
+// selectors must be set.  If a consumer observes none of its fields are set,
+// they must assume an unknown option has been specified and fail closed.
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type PodPeer struct {
+	// PodSelector is a labelSelector used to select Pods, This field is NOT optional,
+	// follows standard label selector semantics and if present but empty, it selects
+	// all Pods.
+	PodSelector *metav1.LabelSelector `json:"podSelector"`
+}
+
+// NamespacedPodPeer defines a flexible way to select Namespaces and pods in a
+// cluster. The `Namespaces` and `PodSelector` fields are required.
+type NamespacedPodPeer struct {
+	// Namespaces is used to select a set of Namespaces.
+	Namespaces NamespacedPeer `json:"namespaces"`
+
+	// Pods is used to select a set of Pods in the set of Namespaces.
+	Pods PodPeer `json:"pods"`
```

Comment:
Why is `PodPeer` needed? Why can't this just be `PodSelector *metav1.LabelSelector`? If we want to define some other way of selecting pods in the future, then that should be a new option under `AdminNetworkPolicyPeer`, not a new option inside `NamespacedPodPeer`.

Reply:
After taking about this in the SIG meeting we decided to remove the `PodPeer` level of indirection